connectors/src/connectors/notion/temporal/activities.ts

```diff
@@ -1984,3 +1968,116 @@ export async function getDiscoveredResourcesFromCache(
     databaseIds: discoveredDatabaseIds,
   };
 }
+
+/** Render page sections according to Notion structure:
+ * - the natural nesting of blocks is used as structure,
+ * - H1 & H2 blocks add a level of nesting in addition to the "natural" nesting,
+ * - only the 2 first levels of nesting are used to create prefixes, similarly
+ *   to github, to avoid too many prefixes
+ */
+function renderPageSection({
+  blocksByParentId,
+}: {
+  blocksByParentId: Record<string, NotionConnectorBlockCacheEntry[]>;
+}): CoreAPIDataSourceDocumentSection {
+  const renderedPageSection: CoreAPIDataSourceDocumentSection = {
+    prefix: null,
+    content: "\n", // add a newline to separate the page from the metadata above (title, author...)
+    sections: [],
+  };
+
+  // Change block parents so that H1/H2 blocks are treated as nesting
+  // for that we need to traverse with a topological sort, leafs treated first
+  const orderedParentIds: string[] = [];
+  const addNode = (nodeId: string) => {
+    const children = blocksByParentId[nodeId];
+    if (!children) return;
+    orderedParentIds.push(nodeId);
+    for (const child of children) {
+      addNode(child.notionBlockId);
+    }
+  };
+  addNode("root");
+  orderedParentIds.reverse();
+
+  const adaptedBlocksByParentId: Record<
+    string,
+    NotionConnectorBlockCacheEntry[]
+  > = {};
+
+  for (const parentId of orderedParentIds) {
+    const blocks = blocksByParentId[
+      parentId
+    ] as NotionConnectorBlockCacheEntry[];
+    blocks.sort((a, b) => a.indexInParent - b.indexInParent);
+    const currentHeadings: { h1: string | null; h2: string | null } = {
+      h1: null,
+      h2: null,
+    };
+    for (const block of blocks) {
+      if (block.blockType === "heading_1") {
+        adaptedBlocksByParentId[parentId] = [
+          ...(adaptedBlocksByParentId[parentId] ?? []),
+          block,
+        ];
+        currentHeadings.h1 = block.notionBlockId;
+        currentHeadings.h2 = null;
+      } else if (block.blockType === "heading_2") {
+        const h2ParentId = currentHeadings.h1 ?? parentId;
+        adaptedBlocksByParentId[h2ParentId] = [
+          ...(adaptedBlocksByParentId[h2ParentId] ?? []),
+          block,
+        ];
+        currentHeadings.h2 = block.notionBlockId;
+      } else {
+        const currentParentId =
+          currentHeadings.h2 ?? currentHeadings.h1 ?? parentId;
+        adaptedBlocksByParentId[currentParentId] = [
+          ...(adaptedBlocksByParentId[currentParentId] ?? []),
+          block,
+        ];
+      }
+    }
+  }
+
+  const renderBlockSection = (
+    b: NotionConnectorBlockCacheEntry,
+    depth: number,
+    indent = ""
+  ): CoreAPIDataSourceDocumentSection => {
+    // Initial rendering (remove base64 images from rendered block)
+    let renderedBlock = b.blockText ? `${indent}${b.blockText}` : "";
+    renderedBlock = renderedBlock
```

Comment:
@fontanierh sanity check: moved image replacement logic from full parsed page to block level, seems ok but wanted to check with you in case it had a specific importance